component/pyroscope/ebpf/ebpf_linux.go

```diff
@@ -0,0 +1,253 @@
+	targetFinder *sd.TargetFinder
+	session      *ebpfspy.Session
+
+	debugInfo     DebugInfo
+	debugInfoLock sync.Mutex
+	metrics       *Metrics
+}
+
+func (c *Component) Run(ctx context.Context) error {
+	err := c.session.Start()
+	if err != nil {
+		return fmt.Errorf("ebpf profiling session start: %w", err)
+	}
+	defer c.session.Stop()
+
+	var g run.Group
+	g.Add(func() error {
+		collectInterval := c.args.CollectInterval
+		t := time.NewTicker(collectInterval)
+		defer t.Stop()
+		for {
+			select {
+			case <-ctx.Done():
+				return nil
+			case newArgs := <-c.argsUpdate:
+				c.args = newArgs
+				c.updateTargetFinder()
+				c.session.UpdateCacheOptions(cacheOptionsFromArgs(c.args))
+				err := c.session.UpdateSampleRate(c.args.SampleRate)
+				if err != nil {
+					return nil
+				}
+				c.appendable.UpdateChildren(newArgs.ForwardTo)
+				if c.args.CollectInterval != collectInterval {
+					t.Reset(c.args.CollectInterval)
+					collectInterval = c.args.CollectInterval
+				}
+			case <-t.C:
+				err := c.CollectProfiles()
+				if err != nil {
+					c.metrics.ProfilingSessionsFailingTotal.Inc()
+					return err
+				}
+				c.updateDebugInfo()
+			}
+		}
+	}, func(error) {
+
+	})
+	return g.Run()
+}
+
+func cacheOptionsFromArgs(args Arguments) symtab.CacheOptions {
+	return symtab.CacheOptions{
+		PidCacheOptions: symtab.GCacheOptions{
+			Size:       args.PidCacheSize,
+			KeepRounds: args.CacheRounds,
+		},
+		BuildIDCacheOptions: symtab.GCacheOptions{
+			Size:       args.BuildIDCacheSize,
+			KeepRounds: args.CacheRounds,
+		},
+		SameFileCacheOptions: symtab.GCacheOptions{
+			Size:       args.SameFileCacheSize,
+			KeepRounds: args.CacheRounds,
+		},
+	}
+}
+
+func (c *Component) updateTargetFinder() {
+	c.targetFinder.SetTargets(sd.TargetsOptions{
+		Targets:       c.args.Targets,
+		DefaultTarget: nil,
+		TargetsOnly:   true,
+	})
+	c.targetFinder.ResizeContainerIDCache(c.args.ContainerIDCacheSize)
+	c.metrics.TargetsActive.Set(float64(len(c.targetFinder.Targets())))
+}
+
+func (c *Component) Update(args component.Arguments) error {
+	newArgs := args.(Arguments)
+	c.argsUpdate <- newArgs
+	return nil
+}
+
+func (c *Component) CollectProfiles() error {
+	c.metrics.ProfilingSessionsTotal.Inc()
+	level.Debug(c.options.Logger).Log("msg", "ebpf  CollectProfiles")
+	args := c.args
+	builders := ebpfspy.NewProfileBuilders(args.SampleRate)
+	err := c.session.CollectProfiles(func(target *sd.Target, stack []string, value uint64, pid uint32) {
+		labelsHash, labels := target.Labels()
+		builder := builders.BuilderForTarget(labelsHash, labels)
+		builder.AddSample(stack, value)
+	})
+
+	if err != nil {
+		return fmt.Errorf("ebpf session CollectProfiles %w", err)
+	}
+	level.Debug(c.options.Logger).Log("msg", "ebpf  CollectProfiles done", "profiles", len(builders.Builders))
+	bytesSent := 0
+	for _, builder := range builders.Builders {
+		c.metrics.PprofsTotal.Inc()
+		var buf bytes.Buffer
+		err := builder.Profile.Write(&buf)
```

Comment:
Looking at `Write` implementation https://github.com/google/pprof/blob/91b7bce49751/profile/profile.go#L341

I think it might be better to use `WriteUncompressed` so that you can pool gzip writers (by resetting) and use this one instead https://pkg.go.dev/github.com/klauspost/compress/gzip#NewWriter

Reply:
Not sure if this will help but worth trying since it ~45% of alloc space 
<img width="648" alt="image" src="https://github.com/grafana/agent/assets/1053421/d2777554-ea42-4372-b5ca-9860b6548428">

Reply:
pooling gzip writers increased inuse mem by ~40M and decreased cpu ~10%. lets keep it pooled and if we ever need to save more memory we could switch back or introduce a configuration option.